src/app/icd/client/RefreshKeySender.h

```diff
@@ -0,0 +1,137 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include "ICDClientInfo.h"
+#include "ICDClientStorage.h"
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/CommandSender.h>
+#include <app/OperationalSessionSetup.h>
+
+#include <crypto/CHIPCryptoPAL.h>
+#include <lib/core/CHIPConfig.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/core/ScopedNodeId.h>
+#include <lib/support/CodeUtils.h>
+#include <stddef.h>
+
+namespace chip {
+namespace app {
+
+class CheckInDelegate;
+/**
+ * @brief RefreshKeySender contains all the data and methods needed for key refresh and re-registration of an ICD client.
+ */
+class RefreshKeySender
+{
+public:
+    typedef Crypto::SensitiveDataBuffer<Crypto::kAES_CCM128_Key_Length> RefreshKeyBuffer;
+    /**
+     * @brief This class comprises command sender callbacks for the registration command.
+     */
+    class Callback : public CommandSender::Callback
+    {
+    public:
+        // CommandSender callbacks
+        void OnResponse(CommandSender * apCommandSender, const ConcreteCommandPath & aPath, const StatusIB & aStatus,
+                        chip::TLV::TLVReader * aData) override;
+        void OnError(const CommandSender * apCommandSender, CHIP_ERROR aError) override;
+        void OnDone(CommandSender * apCommandSender) override;
+
+        void SetRefreshKeySender(RefreshKeySender * apRefreshKeySender) { mpRefreshKeySender = apRefreshKeySender; }
+
+        virtual ~Callback() {}
+
+    private:
+        CHIP_ERROR mError                     = CHIP_NO_ERROR;
+        RefreshKeySender * mpRefreshKeySender = nullptr;
+    };
+
+    RefreshKeySender(CheckInDelegate * apCheckInDelegate, ICDClientInfo & aICDClientInfo, ICDClientStorage * aICDClientStorage);
```

Comment:
Is there a reason the new key is not just a constructor argument?

Is there a reason aICDClientInfo is not a _const_ reference?  Does this constructor mutate it?

Reply:
No, it does not. Updated it to be a const reference.